docs/user-guide/setting-up-ci-cd.rst

```diff
@@ -0,0 +1,207 @@
+:desc: Set up a CI/CD pipeline to ensure that iterative improvements are tested and deployed with minimum manual effort
+
+.. _setting-up-ci-cd:
+
+Setting up CI/CD
+================
+
+Developing a bot looks a little different than developing a traditional
+software application, but that doesn't mean you should abandon software
+development best practices. Setting up CI/CD ensures that incremental updates
+to your bot are improving it, not harming it.
+
+.. contents::
+   :local:
+   :depth: 2
+
+
+Overview
+--------
+
+Continous Integration (CI) is the practice of merging in code changes
+frequently, and automatically testing changes as they are committed. Continuous
+Deployment (CD) means automatically deploying integrated changes to a staging
+or production environment. Together, they allow for more frequent improvements
+to your assistant with less manual effort expended on testing and deployment.
+
+This guide will cover **what** should go in a CI/CD pipeline, specific to a
+Rasa project. **How** you implement that pipeline is up to you. There are many
+CI/CD tools out there, and you might already have a preferred one. Many Git
+repository hosting services like Github, Gitlab, and Bitbucket also provide
+their own CI/CD tools that you can make use of. 
+
+Continuous Integration
+----------------------
+
+Assistants are best improved with frequent, `incremental updates
+<https://rasa.com/docs/rasa-x/user-guide/improve-assistant/>`_.
+No matter how small a change is, you want to be sure that it doesn't introduce
+new problems or negatively impact the performance of your model. Most tests are
+quick enough to run on every change. However, you can set some more
+resource-intentsive tests to run only when certain files have been changed or
+when a certain tag is present.
+
+CI checks usually run on commit, or on merge/pull request.
+
+.. contents::
+   :local:
+
+Validate Data
+#############
+
+:ref:`Data validation <validate-files>` verifies that there are no mistakes or
+major inconsistencies in your domain file, NLU data, or story data. 
+
+.. code-block:: bash
+
+   rasa data validate --fail-on-warnings
+
+If ``rasa
+data validate`` results in errors, training a model will also fail. By
+including the ``--fail-on-warnings`` flag, the command will also fail on
+warnings about problems that won't prevent training a model, but might indicate
+messy data, such as actions listed in the domain that aren't used in any
+stories.
+
+Validate Stories
+################
+
+:ref:`Story validation <test-story-files-for-conflicts>` checks if you have any
+stories where different bot actions follow from the same dialogue history.
+
+.. code-block:: bash
+
+   rasa data validate --fail-on-warnings --max-history 5
+
+Conflicts between stories will prevent Rasa from learning the correct pattern.
+You can run story validation by passing the ``--max-history`` flag to ``rasa
+data validate``, either in a seperate check or as part of the data validation
```

Comment:
What do you mean by this? I guess `rasa data validate` also runs `rasa data validate stories` or no?

Reply:
Not by default, no

Reply:
okay, so what do you mean by "as part of the data validation"?

Reply:
Maybe I shouldn't have said that - I meant, they can just run data validation with the max history flag once and that takes care of both data validation and story validation.

Reply:
ohhh i see so it's the other way around, rasa data validate stories also runs rasa data validate

Reply:
I would probably not mention both. Let's either recommend you add this flag to the previous command or run the whole command again but with the new flag

Reply:
Ok, I think one step makes sense - no point in running the same tests twice right?

Reply:
hmm I guess rasa data validate defaults to including story validation now:
```
$ rasa data validate
2020-04-08 09:48:14 INFO     rasa.validator  - Validating intents...
2020-04-08 09:48:14 INFO     rasa.validator  - Validating uniqueness of intents and stories...
2020-04-08 09:48:14 INFO     rasa.validator  - Validating utterances...
2020-04-08 09:48:14 INFO     rasa.validator  - Story structure validation...
```

Reply:
maybe it makes sense just to merge it into 1 step and describe the 2 things it will do?